cassandra/concurrent.py

```diff
@@ -74,32 +84,121 @@ def execute_concurrent(session, statements_and_parameters, concurrency=100, rais
+    return executor.execute(concurrency, raise_on_first_error)
+
+
+class _ConcurrentExecutor(object):
+
+    def __init__(self, session, statements_and_params):
+        self.session = session
+        self._enum_statements = enumerate(iter(statements_and_params))
+        self._condition = Condition()
+        self._fail_fast = False
+        self._results_queue = []
+        self._current = 0
+        self._exec_count = 0
+
+    def execute(self, concurrency, fail_fast):
+        self._fail_fast = fail_fast
+        self._results_queue = []
+        self._current = 0
+        self._exec_count = 0
+        with self._condition:
+            for n in xrange(concurrency):
+                if not self._execute_next():
+                    break
+        return self._results()
+
+    def _execute_next(self):
+        # lock must be held
+        try:
+            (idx, (statement, params)) = next(self._enum_statements)
+            self._exec_count += 1
+            self._execute(idx, statement, params)
+            return True
+        except StopIteration:
+            pass
+
+    def _execute(self, idx, statement, params):
+        try:
+            future = self.session.execute_async(statement, params, timeout=None)
+            args = (future, idx)
+            future.add_callbacks(
+                callback=self._on_success, callback_args=args,
+                errback=self._on_error, errback_args=args)
+        except Exception as exc:
+            # exc_info with fail_fast to preserve stack trace info when raising on the client thread
+            # (matches previous behavior -- not sure why we wouldn't want stack trace in the other case)
+            e = sys.exc_info() if self._fail_fast and six.PY2 else exc
+            self._put_result(e, idx, False)
+
+    def _on_success(self, result, future, idx):
+        if future.has_more_pages:
+            result = PagedResult(future, result)
+            future.clear_callbacks()
+        self._put_result(result, idx, True)
+
+    def _on_error(self, result, future, idx):
+        self._put_result(result, idx, False)
+
+    @staticmethod
+    def _raise(exc):
         if six.PY2 and isinstance(exc, tuple):
             (exc_type, value, traceback) = exc
             six.reraise(exc_type, value, traceback)
         else:
             raise exc
-    else:
-        return results
+
+
+class ConcurrentExecutorGenResults(_ConcurrentExecutor):
+
+    def _put_result(self, result, idx, success):
+        with self._condition:
+            heappush(self._results_queue, (idx, (success, result)))
+            self._execute_next()
+            self._condition.notify()
+
+    def _results(self):
+        with self._condition:
+            while self._current < self._exec_count:
+                while not self._results_queue or self._results_queue[0][0] != self._current:
```

Comment:
I don't think the `self._results_queue[0][0] != self._current` and `self._results_queue[0][0] == self._current` checks are necessary. Because the result queue is a heap if it has items then the front will alway be current. If that's the case the code could look something like this:

```python
class ConcurrentExecutorGenResults(_ConcurrentExecutor):

    # ...

    def _results(self):
        with self._condition:
            while self._current < self._exec_count:
                if self._results_queue:
                    _, res = heappop(self._results_queue)
                    if self._fail_fast and not res[0]:
                        self._raise(res[1])
                    yield res
                    self._current += 1
                else:
                    self._condition.wait()
```

Or if you want the inner results loop to be a bit tighter you could do something like this (uglier though):

```python
class ConcurrentExecutorGenResults(_ConcurrentExecutor):

    # ...

    def _results(self):
        with self._condition:
            while True:
                while self._results_queue:
                    _, res = heappop(self._results_queue)
                    if self._fail_fast and not res[0]:
                        self._raise(res[1])
                    yield res
                    self._current += 1

                if self._current < self._exec_count:
                    self._condition.wait()
                else:
                    break
```